### PR TITLE
[NFC] Simplify the tiling implementation using cloning.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
The current implementation of tiling using `scf.for` is convoluted to make sure that the destination passing style of the untiled program is preserved. The addition of support to tile using `scf.forall` (adapted from the transform operation in Linalg) in https://github.com/llvm/llvm-project/pull/67083 used cloning of the tiled operations to better streamline the implementation. This PR adapts the other tiling methods to use a similar approach, making the transformations (and handling destination passing style semantics) more systematic.

### CODE REVIEW

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 1]
```diff
@@ -139,15 +142,17 @@ static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
...
+///   the tile processed within the inner most loop.
+/// Note that this methods adds `scf.yield` operation for all but the innermost
+/// loop. These yield the value returned by the immediately inner loop. The
+/// caller is expected to add the scf.yield operation for the innermost loop.
 static SmallVector<scf::ForOp> generateTileLoopNest(
     OpBuilder &builder, Location loc, ArrayRef<Range> loopRanges,
     ArrayRef<OpFoldResult> tileSizes, SmallVector<OpFoldResult> &offsets,
-    SmallVector<OpFoldResult> &sizes) {
-  assert(!loopRanges.empty() && "expected at least one loop range");
-  assert(loopRanges.size() == tileSizes.size() &&
```
Comment (matthias-springer) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:149]: why was this deleted?

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 2]
```diff
@@ -128,6 +128,9 @@ static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
                                                   Operation *op,
                                                   ValueRange newDestArgs) {
   Operation *clonedOp = rewriter.clone(*op);
+  if (newDestArgs.empty()) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:131]: nit: trivial braces can be omitted

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 3]
```diff
@@ -169,136 +174,98 @@ static SmallVector<scf::ForOp> generateTileLoopNest(
     }
 
     auto loop = builder.create<scf::ForOp>(
```
Comment (matthias-springer) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:176]: Maybe `mlir::scf::buildLoopNest` can be used to create the loop nest.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 4]
```diff
@@ -169,136 +174,98 @@ static SmallVector<scf::ForOp> generateTileLoopNest(
...
-    auto sliceOp = destValue.value().getDefiningOp<tensor::ExtractSliceOp>();
-    if (!sliceOp)
-      continue;
-    sliceOp.setOperand(0, bbArgsList[destValue.index()]);
+/// Method to add new init values to a loop nest. Updates `loops` in-place with
+/// new loops that use the `newInitValues`.
+/// The outer-loops are updated to yield the new result values of the inner
+/// loop. For the innermost loop, the call back `getNewYields` is invoked to get
+/// the additional values to yield form the innermost loop.
+static void addInitOperandsToLoopNest(
```
Comment (matthias-springer) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:205]: I think this can be simplified by calling `LoopLikeOpInterface::replaceWithAdditionalYields` for each loop. `replaceWithAdditionalYields` does not give you the IV in the NewValuesFn lambda, but it's also not needed here (`newYieldValuesFn`). (If you need it at some point, we can have the lambda provide the new loop op, from which it can be queried.)

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 5]
```diff
@@ -169,136 +174,98 @@ static SmallVector<scf::ForOp> generateTileLoopNest(
...
+    RewriterBase &rewriter, MutableArrayRef<scf::ForOp> loops,
+    ValueRange newInitValues,
+    llvm::function_ref<SmallVector<Value>(RewriterBase &rewriter, Value iv,
+                                          ValueRange newRegionIterArgs)>
+        getNewYieldValsFn) {
+  SmallVector<scf::ForOp> newLoops;
+  if (loops.empty()) {
```
Comment (chelini) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:212]: nit: trivial braces can be omitted

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 6]
```diff
@@ -544,17 +554,56 @@ mlir::scf::tileAndFuseProducerOfSlice(RewriterBase &rewriter,
...
+  Operation *clonedProducerOp = cloneOpAndUpdateDestinationArgs(
+      rewriter, fusableProducerOp, clonedOpDestinationTensors);
+  // 2d. Update the source of the candidateSlice to be the cloned producer.
+  //     Easier to just clone the slice with different source since replacements
+  //     and DCE of cloned ops becomes easier
+  SmallVector<Value> candidateSliceOpOperands =
+      llvm::to_vector(candidateSliceOp->getOperands());
+  candidateSliceOpOperands[0] = clonedProducerOp->getResult(resultNumber);
+  tensor::ExtractSliceOp clonedCandidateSliceOp =
+      mlir::clone(rewriter, candidateSliceOp,
```
Comment (chelini) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:591]: Is this a good way of cloning the op? My understanding was always to use the builder to signal the listener about IR mutation (@matthias-springer).

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 7]
```diff
@@ -169,136 +174,98 @@ static SmallVector<scf::ForOp> generateTileLoopNest(
...
+  for (auto &loop : loops) {
+    rewriter.setInsertionPoint(loop);
+
+    // Create a new loop with the new init values for this loop.
+    SmallVector<Value> newInits = llvm::to_vector(loop.getInitArgs());
+    newInits.append(newInitValues.begin(), newInitValues.end());
+    auto newLoop = rewriter.create<scf::ForOp>(
+        loop.getLoc(), loop.getLowerBound(), loop.getUpperBound(),
+        loop.getStep(), newInits,
+        [&](OpBuilder &b, Location loc, Value iv, ValueRange iterArgs) {});
```
Comment (chelini) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:226]: I believe we can avoid the lambda here.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 8]
```diff
@@ -169,136 +174,98 @@ static SmallVector<scf::ForOp> generateTileLoopNest(
...
+                        innerMostLoop.getRegionIterArgs());
+  SmallVector<Value> newYieldOperands =
+      llvm::to_vector(innerMostYieldOp->getOperands());
+  newYieldOperands.append(newYieldVals);
+  rewriter.replaceOpWithNewOp<scf::YieldOp>(innerMostYieldOp, newYieldOperands);
+
+  // Make all other loops except the innermost loops yield the values returned
+  // by the inner loop.
+  for (auto [outerLoop, innerLoop] :
+       llvm::zip(loops.drop_back(), loops.drop_front())) {
```
Comment (chelini) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:258]: llvm::zip_equal?

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 9]
```diff
@@ -317,10 +284,6 @@ mlir::scf::tileUsingSCFForOp(RewriterBase &rewriter, TilingInterface op,
   // 1. Get the range of the loops that are represented by the operation.
   SmallVector<Range> iterationDomain = op.getIterationDomain(rewriter);
   size_t numLoops = iterationDomain.size();
-  if (numLoops == 0) {
```
Comment (chelini) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:320]: why did you remove it?

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 10]
```diff
@@ -636,28 +664,46 @@ void mlir::scf::yieldReplacementForFusedProducer(
     RewriterBase &rewriter, tensor::ExtractSliceOp sliceOp,
     scf::SCFFuseProducerOfSliceResult fusedProducerInfo,
     MutableArrayRef<scf::ForOp> loops) {
-  auto [fusableProducer, fusedProducerValue, tileAndFusedOps] =
-      fusedProducerInfo;
-  SmallVector<Value> initValues;
+  if (loops.empty()) {
```
Comment (chelini) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:667]: nit: trivial braces can be omitted

File: mlir/test/Dialect/Tensor/tiling.mlir [hunk 11]
```diff
@@ -374,8 +374,8 @@ module attributes {transform.with_named_sequence} {
 // CHECK:               %[[IN_J:.*]] = affine.apply #[[MAP2]](%[[J]])[%[[TILE_1]]]
 // CHECK:               %[[IN_J_SZ:.*]] = affine.min #[[MAP3]](%[[OUT_J_SZ]], %[[J]])[%[[TILE_1]], %[[IN_D1]]]
 // CHECK:               %[[SUB_IN:.*]] = tensor.extract_slice %[[IN]][%[[IN_I]], %[[IN_J]]] [%[[IN_I_SZ]], %[[IN_J_SZ]]] [1, 1] : tensor<?x?xf32> to tensor<?x?xf32>
-// CHECK:               %[[OUT_D2:.+]] = tensor.dim %[[OUT]], %[[C2]]
```
Comment (chelini) [inline mlir/test/Dialect/Tensor/tiling.mlir:377]: Can you please comment on this change?

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 12]
```diff
@@ -169,17 +174,25 @@ static SmallVector<scf::ForOp> generateTileLoopNest(
...
     offsets[loopRange.index()] = loop.getInductionVar();
     loops.push_back(loop);
-    builder.setInsertionPoint(loop.getBody()->getTerminator());
+    builder.setInsertionPointToEnd(loop.getBody());
+    destinationTensors = loop.getRegionIterArgs();
+  }
+
+  // Add the scf.yield operations for all the outer loops.
+  for (auto [outerLoop, innerLoop] :
+       llvm::zip(MutableArrayRef(loops).drop_back(),
```
Comment (hanhanW) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:192]: nit: use `zip_equal`. It has an assertion check, which looks better to me.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 13]
```diff
@@ -583,17 +583,54 @@ mlir::scf::tileAndFuseProducerOfSlice(RewriterBase &rewriter,
...
   rewriter.setInsertionPoint(candidateSliceOp);
+
+  // 2. Clone the fused producer
+  // 2a. Compute the destination operands to use for the cloned operation.
+  SmallVector<Value> origDestinationTensors, clonedOpDestinationTensors;
+  Operation *fusableProducerOp = fusableProducer.getOwner();
+  if (isa<DestinationStyleOpInterface>(fusableProducerOp)) {
+    if (failed(tensor::getOrCreateDestinations(
+            rewriter, fusableProducerOp->getLoc(), fusableProducerOp,
+            origDestinationTensors))) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:569]: maybe we can just merge them into a single `if`?

**Comment by MaheshRavishankar** [comment]
> Quick offhand question before I can devote more time to a review later this week: is this a step towards reducing the duplicated implementation problem ?

Yes it is (at least in my mind)... The reasons I did this was after looking at the implementation of the tiling using `scf.forall` under the hood of the transform dialect op, I realized that I could adapt the current implementation of tiling using `scf.for` much simpler by following the same path. So before I push forward on moving more of the implementation of tiling from under the transform dialect op in Linalg into here, took a slight detour to make the current implementation of `scf.for` consistent with the `scf.forall` implementation. I think there is a lot of logic here we could reuse for both.

In terms of taking time, yeah, I still need to test it with IREE and fix any issues. So its not urgent, but quicker review is appreciated. 

PS: irrespective of the above, I think this is just in general a good cleanup. It handles better managing the destination passing style through tiling.

**Review by nicolasvasilache — COMMENTED** [review]
I remember being quite unsatisfied with the genuflexions related to lack of scoping in `updateDestinationOperandsForTiledOp` and other functions introduced in https://reviews.llvm.org/D134144.

It is nice to see all this go, thanks for the cleanup!

Waiting on understanding the trajectory of the commit that I skipped to approve.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 14]
```diff
@@ -708,28 +781,46 @@ void mlir::scf::yieldReplacementForFusedProducer(
     RewriterBase &rewriter, tensor::ExtractSliceOp sliceOp,
     scf::SCFFuseProducerOfSliceResult fusedProducerInfo,
     MutableArrayRef<scf::ForOp> loops) {
-  auto [fusableProducer, fusedProducerValue, tileAndFusedOps] =
-      fusedProducerInfo;
-  SmallVector<Value> initValues;
+  if (loops.empty()) {
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp]: > None of this code is actually used in MLIR besides a test.
> 
> Is there a plan to use this in some other implementation (e.g. fusion in LinalgTransformOps) or are we in another occurrence of multiple implementations?
> 
> In fact, I see in `TestTilingInterface::TestTileConsumerFuseAndYieldProducerUsingSCFForOp` that:
> 
> ```
>     // The rest of this method is similar to
>     // scf::tileConsumerAndFuseProducerGreedilyUsingSCFForOp, except that also
>     // yields replacements for values of the fused producer.
> ```
> 
> Why do we need this / should we replace `tileConsumerAndFuseProducerGreedilyUsingSCFForOp` by this (it seems to have more use cases) ?
> 
> Skipping the review of this commit until we get clarification on the intended direction.

Yes, I agree this part of the code is awkward and this is not what I would do either. I think this was just a working compromise. I wanted to fix that up as well, ill describe below how, but my aim with this PR is too remove the `yieldTiledValues` and `updateDestinationOperandsForTiledOp` which never really worked well, and I figured how to fix that only after seeing it work better using the cloning approach. So for now, I'd rather land this PR as an NFC (which it is, is there something that stands out to you as not being NFC, cause that is not the intent)

Coming back to this method, I would rather remove the somewhat artificially split up that introduces `tileAndFuseProducerOfSlice`  and `yieldReplacementForFusedProducer`. (FWIW I guess the split isnt too bad cause I changed the implementation underneath without having to change these methods, which is something). But I want to support the use case that is represented by the test here https://github.com/llvm/llvm-project/blob/main/mlir/test/Interfaces/TilingInterface/tile-fuse-and-yield-using-interface.mlir . Without going into details, the intent here is that when you tile and fuse operations, in some cases you also want to return a replacement for the fused producer (just like the tiling returns a replacement for the original untiled operation). Reason it is important is consider a function that is returning both the consumer op and the producer op. If you are tiling the consumer and fusing the producer with it, if you dont return a replacement value for the producer, the original producer stays live (and that actually defeats the purpose of the fusion).

So my proposed fix up (as a follow up), is to allow an extra list of values passed as arguments to `tileConsumerAndFuseProducerGreedily..` . If this value is tiled and fused , the loop will return a replacement value for it. I can send out a draft PR for it on top of this if this helps, but I hope we can decouple the two. This PR should be purely NFC.

**Comment by MaheshRavishankar** [comment]
@nicolasvasilache , @matthias-springer and @chelini This is ready to land. After this I can start working on unifying the different tiling methods.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp]: SG, I appreciate the removal of `yieldTiledValues` and `updateDestinationOperandsForTiledOp`

**Review by nicolasvasilache — APPROVED** [review]

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 6 above)
Comment (chelini) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:592]: Just FYI, `mlir::clone` does not notify the cloned operations in the op region (https://github.com/chelini/tpp-sandbox/commit/e6d833b8ae1bf7f820a778d2bf7517f9b03e8b97). but here is not an issue.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 15]
```diff
@@ -636,28 +665,46 @@ void mlir::scf::yieldReplacementForFusedProducer(
...
+
+        Block *block = rewriter.getInsertionPoint()->getBlock();
+        rewriter.setInsertionPoint(block->getTerminator());
+        Value replacement = rewriter.create<tensor::InsertSliceOp>(
+            fusedProducerInfo.origProducer.getLoc(),
+            fusedProducerInfo.tiledAndFusedProducer,
+            loops.back().getRegionIterArgs().back(), sliceOp.getMixedOffsets(),
+            sliceOp.getMixedSizes(), sliceOp.getMixedStrides());
+        return {replacement};
+      }
```
Comment (tlongeri) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:702]: It looks like you're missing a return value for when the if condition is false? I am getting build errors over this.

Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:702]: ah, sorry. ill fix that pronto
